File: ideal-core/src/main/java/cn/idealframework2/idempotent/memory/CaffeineIdempotentHandlerFactory.java
package cn.idealframework2.idempotent.memory;

import cn.idealframework2.idempotent.IdempotentHandler;
import cn.idealframework2.idempotent.IdempotentHandlerFactory;
import com.github.benmanes.caffeine.cache.Cache;

import javax.annotation.Nonnull;
import java.time.Duration;

/**
 * @author 宋志宗 on 2022/12/6
 */
public class CaffeineIdempotentHandlerFactory implements IdempotentHandlerFactory {

  @Nonnull
  @Override
  public IdempotentHandler create(@Nonnull String namespace, @Nonnull Duration expire) {
    long millis = expire.toMillis();
    Cache<String, Boolean> cache = CaffeineIdempotentCacheFactory.get(millis);
    return new CaffeineIdempotentHandler(namespace, cache);
  }
}

File: ideal-core/src/main/kotlin/cn/idealframework2/idempotent/coroutine/memory/CaffeineIdempotentHandlerFactory.kt
package cn.idealframework2.idempotent.coroutine.memory

import cn.idealframework2.idempotent.coroutine.IdempotentHandler
import cn.idealframework2.idempotent.coroutine.IdempotentHandlerFactory
import cn.idealframework2.idempotent.memory.CaffeineIdempotentCacheFactory
import java.time.Duration

/**
 * @author 宋志宗 on 2022/12/6
 */
class CaffeineIdempotentHandlerFactory : IdempotentHandlerFactory {

  override fun create(namespace: String, expire: Duration): IdempotentHandler {
    val millis = expire.toMillis()
    val cache = CaffeineIdempotentCacheFactory.get(millis)
    return CaffeineIdempotentHandler(namespace, cache)
  }
}
